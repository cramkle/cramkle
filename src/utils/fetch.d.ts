import { fetch } from 'typescript/lib/dom'

export default fetch
