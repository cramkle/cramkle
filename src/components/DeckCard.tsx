import { plural } from '@lingui/macro'
import { useLingui } from '@lingui/react'
import gql from 'graphql-tag'
import React from 'react'
import { useHistory } from 'react-router'

import { FlashCardStatus } from '../globalTypes'
import FlashCardStatusChip from './FlashCardStatus'
import { DeckCard_deck } from './__generated__/DeckCard_deck'
import { Card, CardPressable } from './views/Card'
import { Body2, Headline3 } from './views/Typography'

interface Props {
  deck: DeckCard_deck
  onClick?: (deck: DeckCard_deck) => void
  showStudySessionDetails?: boolean
}

export const deckCardFragment = gql`
  fragment DeckCard_deck on Deck {
    id
    slug
    title
    description
    totalNotes
    totalFlashcards
    studySessionDetails {
      newCount
      learningCount
      reviewCount
    }
  }
`

const DeckCard: React.FunctionComponent<Props> = ({
  onClick,
  deck,
  showStudySessionDetails = false,
}) => {
  const history = useHistory()
  const { i18n } = useLingui()

  const handleClick = () => {
    if (onClick) {
      onClick(deck)
    } else {
      history.push(`/d/${deck.slug}`)
    }
  }

  const {
    studySessionDetails: { newCount, learningCount, reviewCount },
  } = deck

  const uniqueId = `deck-${deck.id}`

  return (
    <Card lean className="h-full">
      <CardPressable
        className="p-2 h-full"
        tabIndex={0}
        role="article"
        onClick={handleClick}
        onKeyDown={(e: React.KeyboardEvent) =>
          e.key === 'Enter' && handleClick()
        }
        aria-describedby={`${uniqueId}-title`}
      >
        <Headline3 id={`${uniqueId}-title`}>{deck.title}</Headline3>
        {deck.description && <Body2>{deck.description}</Body2>}
        {showStudySessionDetails ? (
          <div className="mt-1 flex">
            {newCount > 0 && (
              <FlashCardStatusChip
                status={FlashCardStatus.NEW}
                className="mr-2"
                truncated
              >
                {i18n._(plural(newCount, { one: '# new', other: '# new' }))}
              </FlashCardStatusChip>
            )}
            {learningCount > 0 && (
              <FlashCardStatusChip
                status={FlashCardStatus.LEARNING}
                className="mr-2"
                truncated
              >
                {i18n._(
                  plural(learningCount, {
                    one: '# learning',
                    other: '# learning',
                  })
                )}
              </FlashCardStatusChip>
            )}
            {reviewCount > 0 && (
              <FlashCardStatusChip status={FlashCardStatus.REVIEW} truncated>
                {i18n._(
                  plural(reviewCount, { one: '# review', other: '# review' })
                )}
              </FlashCardStatusChip>
            )}
          </div>
        ) : (
          <Body2 className="flex items-center">
            {i18n._(
              plural(deck.totalNotes, {
                one: '# note',
                other: '# notes',
              })
            )}
            <span className="inline-block mx-1">&middot;</span>
            {i18n._(
              plural(deck.totalFlashcards, {
                one: '# flashcard',
                other: '# flashcards',
              })
            )}
          </Body2>
        )}
      </CardPressable>
    </Card>
  )
}

export default DeckCard
