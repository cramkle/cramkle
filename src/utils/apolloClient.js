import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { HttpLink } from 'apollo-link-http'
import { onError } from 'apollo-link-error'
import { ApolloLink } from 'apollo-link'
import { canUseDOM } from 'exenv'

export const createApolloClient = uri => {
  const cache = new InMemoryCache()

  const client = new ApolloClient({
    ssrMode: process.env.SSR,
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors) {
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message ${message}, Location: ${locations}, Path ${path}`
            )
          )
        }
        if (networkError) {
          console.log(`[Network error]: ${networkError}`)
        }
      }),
      new HttpLink({
        uri,
        credentials: 'include',
        fetch: window.fetch,
      }),
    ]),
    // eslint-disable-next-line no-underscore-dangle
    cache: canUseDOM ? cache.restore(window.__APOLLO_STATE__) : cache,
  })

  return client
}

const defaultClient = createApolloClient('http://localhost:5000/graphql')

export default defaultClient
