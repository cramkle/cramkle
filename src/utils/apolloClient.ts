import { ApolloClient, ApolloLink, createHttpLink } from '@apollo/client'
import { InMemoryCache } from '@apollo/client/cache'
import type { NormalizedCacheObject } from '@apollo/client/cache'
import { onError } from '@apollo/client/link/error'

declare global {
  interface Window {
    __APOLLO_STATE__: NormalizedCacheObject
  }

  namespace NodeJS {
    interface Process {
      browser: boolean
    }
  }
}

export const createApolloClient = (uri: string, cookie?: string) => {
  const cache = new InMemoryCache()

  const errorLink = onError(({ graphQLErrors, networkError }) => {
    if (graphQLErrors) {
      graphQLErrors.forEach(({ message, locations, path }) =>
        console.error(
          `[GraphQL error]: Message ${message}, Location: ${locations}, Path ${path}`
        )
      )
    }
    if (networkError) {
      console.error(`[Network error]: ${networkError}`)
    }
  })

  const httpLink = createHttpLink({
    uri,
    credentials: 'include',
    fetch,
    headers: {
      cookie,
    },
    fetchOptions: {
      method: 'POST',
    },
  })

  const client = new ApolloClient({
    ssrMode: !process.browser,
    link: ApolloLink.from([errorLink, httpLink]),
    cache:
      typeof window !== 'undefined'
        ? cache.restore(window.__APOLLO_STATE__)
        : cache,
  })

  return client
}
