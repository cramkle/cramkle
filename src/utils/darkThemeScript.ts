export const darkThemeHelmetScript = (darkMode?: boolean) => {
  return `
 (function() {
  window.__onThemeChange = function() {};
  function setTheme(newTheme) {
    window.__theme = newTheme;
    preferredTheme = newTheme;
    document.documentElement.classList.remove('__' + (newTheme === 'dark' ? 'light': 'dark') + '-mode');
    document.documentElement.classList.add('__' + newTheme + '-mode');
    window.__onThemeChange(newTheme);
  }

  var preferredTheme = ${JSON.stringify(
    darkMode == null ? undefined : darkMode ? 'dark' : 'light'
  )};

  window.__setPreferredTheme = function(newTheme) {
    setTheme(newTheme);
    try {
      localStorage.setItem('theme', newTheme);
    } catch (err) {}
  }

  setTheme(preferredTheme || 'light');
})();
`.trim()
}
