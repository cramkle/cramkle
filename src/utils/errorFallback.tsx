import { Trans } from '@lingui/macro'

import { Button } from '../components/views/Button'
import { Body1, Headline4, Overline } from '../components/views/Typography'

export const errorFallback = ({
  error,
  componentStack,
}: {
  error: Error
  componentStack: string | null
}) => {
  return (
    <div className="flex flex-col justify-center items-center py-8 min-h-screen bg-background  bg-opacity-background">
      <Headline4 className="text-txt text-opacity-text-primary">
        <Trans>Something went wrong</Trans>
      </Headline4>
      <Body1 className="mt-1 text-txt text-opacity-text-primary">
        <Trans>Please, try reloading the page</Trans>
      </Body1>
      <Button
        id="refresh-button"
        className="mt-4"
        onClick={() => window.location.reload()}
      >
        <Trans>Refresh</Trans>
      </Button>
      {process.env.NODE_ENV === 'development' && error && (
        <>
          <Overline className="mt-4 text-txt text-opacity-text-primary">
            Error message: "{error.message}"
          </Overline>
          <pre className="text-txt text-opacity-text-primary">
            {componentStack ?? error.stack}
          </pre>
        </>
      )}
    </div>
  )
}
