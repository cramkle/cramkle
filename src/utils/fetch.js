if (process.env.SSR) {
  module.exports = require('node-fetch/lib/index.js')
} else {
  module.exports = window.fetch
}
