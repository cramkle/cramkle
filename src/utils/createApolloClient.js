import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { HttpLink } from 'apollo-link-http'
import { onError } from 'apollo-link-error'
import { ApolloLink } from 'apollo-link'

import fetch from './fetch'

const createApolloClient = uri => {
  const client = new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors) {
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message ${message}, Location: ${locations}, Path ${path}`
            )
          )
        }
        if (networkError) {
          console.log(`[Network error]: ${networkError}`)
        }
      }),
      new HttpLink({
        uri,
        credentials: 'same-origin',
        fetch,
      }),
    ]),
    cache: new InMemoryCache(),
  })

  return client
}

export default createApolloClient
