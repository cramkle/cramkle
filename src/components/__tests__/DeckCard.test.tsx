import React from 'react'

import { render } from '../../test/utils'
import DeckCard from '../DeckCard'

describe('<DeckCard />', () => {
  it('should match snapshot', () => {
    const comp = render(
      <DeckCard
        deck={{
          __typename: 'Deck',
          id: '123',
          slug: 'nihongo-genki',
          title: '日本語げんき',
          description: '',
          totalNotes: 2000,
          totalFlashcards: 3200,
          studySessionDetails: {
            __typename: 'StudySessionDetails',
            newCount: 0,
            learningCount: 0,
            reviewCount: 0,
          },
        }}
      />
    )

    expect(comp.container).toMatchSnapshot()
  })
})
