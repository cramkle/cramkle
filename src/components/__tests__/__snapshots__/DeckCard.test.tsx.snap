// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeckCard /> should match snapshot 1`] = `
<div>
  <div
    class="bg-surface text-on-surface rounded border border-gray-1 h-full"
  >
    <div
      aria-describedby="deck-123-title"
      class="flex flex-col relative outline-none cursor-pointer p-2 h-full"
      role="article"
      tabindex="0"
    >
      <h3
        class="text-xl leading-normal font-medium"
        id="deck-123-title"
      >
        日本語げんき
      </h3>
      
      <p
        class="flex items-center text-secondary text-sm"
      >
        2,000 notes
        <span
          class="inline-block mx-1"
        >
          ·
        </span>
        3,200 flashcards
      </p>
      <div
        class="absolute z-1 top-0 left-0 right-0 bottom-0 bg-hover-overlay transition-opacity ease-in-out duration-200 opacity-0"
      />
    </div>
  </div>
</div>
`;
